File: apps/api/src/rules/geofence-rules.ts
import { DepotSchedule, isAfterReturnTime } from './schedule';

/**
 * Règles métier du geofencing, côté serveur — cf. docs/06 §4.2 et §5.
 *
 * Le serveur ne réévalue PAS la géométrie : l'appareil lui transmet des
 * transitions déjà confirmées par son moteur local (hystérésis, échantillons
 * multiples, précision). Le serveur applique la règle horaire avec SA propre
 * horloge et la configuration du dépôt à jour, et c'est lui qui fait autorité
 * en cas de divergence.
 *
 * Fonctions pures : entrée -> décision. Aucun accès base, aucun effet de bord.
 */

export type TransitionKind = 'ENTER' | 'EXIT';

export type GeofenceEventTypeName =
  | 'ENTER_DEPOT'
  | 'EXIT_DEPOT'
  | 'ENTER_DEPOT_AFTER_RETURN_TIME'
  | 'AFTER_RETURN_EXIT';

export type SessionStateName = 'ACTIVE' | 'RETURNED';

export interface GeofenceDecisionInput {
  transition: TransitionKind;
  /** Instant de la transition, tel que daté par l'appareil (UTC). */
  occurredAt: Date;
  schedule: DepotSchedule;
  /** État de la session au moment de la transition. */
  sessionState: SessionStateName;
}

export interface GeofenceDecision {
  eventType: GeofenceEventTypeName;
  nextSessionState: SessionStateName;
  /** Le téléphone doit-il être marqué comme retourné (horodatage + position) ? */
  markReturned: boolean;
  alert: {
    type: 'AFTER_RETURN_EXIT';
    severity: 'HIGH';
    title: string;
  } | null;
  /** Explication lisible, conservée dans l'événement pour le support. */
  reason: string;
}

export function evaluateGeofenceTransition(
  input: GeofenceDecisionInput,
): GeofenceDecision {
  const { transition, occurredAt, schedule, sessionState } = input;

  if (transition === 'ENTER') {
    const afterReturn = isAfterReturnTime(schedule, occurredAt);

    if (!afterReturn) {
      // Le chauffeur passe au dépôt avant l'heure de retour : c'est un simple
      // passage, pas une fin de tournée.
      return {
        eventType: 'ENTER_DEPOT',
        nextSessionState: sessionState,
        markReturned: false,
        alert: null,
        reason: "Entrée au dépôt avant l'heure de retour : aucun effet.",
      };
    }

    return {
      eventType: 'ENTER_DEPOT_AFTER_RETURN_TIME',
      nextSessionState: 'RETURNED',
      // On ne réécrit pas l'horodatage de retour si la session est déjà
      // marquée retournée : le premier retour fait foi.
      markReturned: sessionState !== 'RETURNED',
      alert: null,
      reason:
        "Entrée au dépôt après l'heure de retour : téléphone considéré comme retourné.",
    };
  }

  // transition === 'EXIT'
  if (sessionState === 'RETURNED') {
    return {
      eventType: 'AFTER_RETURN_EXIT',
      // L'alerte n'est pas un état bloquant : la session redevient ACTIVE,
      // et l'alerte reste ouverte jusqu'à acquittement (cf. docs/02 §6).
      nextSessionState: 'ACTIVE',
      markReturned: false,
      alert: {
        type: 'AFTER_RETURN_EXIT',
        severity: 'HIGH',
        title: 'Sortie du dépôt après le retour',
      },
      reason:
        'Sortie du dépôt alors que le téléphone était marqué comme retourné.',
    };
  }

  return {
    eventType: 'EXIT_DEPOT',
    nextSessionState: sessionState,
    markReturned: false,
    alert: null,
    reason: 'Sortie du dépôt en cours de tournée : aucun effet.',
  };
}

// ---------------------------------------------------------------------------
//  Classification géométrique (miroir du moteur Android)
//
//  Le serveur ne s'en sert pas pour décider, mais pour vérifier la cohérence
//  d'une transition reçue et pour alimenter les scénarios de test partagés.
// ---------------------------------------------------------------------------

export type FixClassification = 'INSIDE_CERTAIN' | 'OUTSIDE_CERTAIN' | 'UNDETERMINED';

const EARTH_RADIUS_M = 6371008.8;

/** Distance haversine en mètres entre deux positions géographiques. */
export function haversineMeters(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number,
): number {
  const toRad = (d: number) => (d * Math.PI) / 180;
  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLon / 2) ** 2;
  return 2 * EARTH_RADIUS_M * Math.asin(Math.min(1, Math.sqrt(a)));
}

/**
 * Classe un point par rapport à une zone, en tenant compte de son incertitude.
 *
 *   d + a  <  R        -> certainement dedans
 *   d - a  >  R + H    -> certainement dehors
 *   sinon              -> indéterminé, AUCUNE transition
 *
 * C'est cette troisième branche qui empêche le faux positif décrit dans la
 * spécification : 50 m hors zone avec 100 m de précision reste indéterminé.
 */
export function classifyFix(params: {
  distanceMeters: number;
  accuracyMeters: number;
  radiusMeters: number;
  hysteresisMeters: number;
}): FixClassification {
  const { distanceMeters: d, accuracyMeters: a, radiusMeters: r, hysteresisMeters: h } =
    params;

  if (d + a < r) return 'INSIDE_CERTAIN';
  if (d - a > r + h) return 'OUTSIDE_CERTAIN';
  return 'UNDETERMINED';
}

File: apps/android/core-rules/src/main/kotlin/com/phonecontrol/core/rules/GeofenceRules.kt
package com.phonecontrol.core.rules

import java.time.Instant
import kotlin.math.PI
import kotlin.math.asin
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * Décisions métier sur une transition de geofence.
 *
 * Portage fidèle de `apps/api/src/rules/geofence-rules.ts`, contraint par les
 * mêmes scénarios (`packages/state-machine-spec/scenarios/depot-rules.json`).
 *
 * Le téléphone applique ces règles pour réagir immédiatement, y compris hors
 * ligne. Le serveur les réapplique à la réception et **fait autorité** : c'est
 * lui qui détient l'heure de référence et la configuration à jour du dépôt.
 */

enum class TransitionKind { ENTER, EXIT }

enum class GeofenceEventType {
    ENTER_DEPOT,
    EXIT_DEPOT,
    ENTER_DEPOT_AFTER_RETURN_TIME,
    AFTER_RETURN_EXIT,
}

enum class SessionState { ACTIVE, RETURNED }

data class GeofenceAlert(
    val type: String = "AFTER_RETURN_EXIT",
    val severity: String = "HIGH",
    val title: String = "Sortie du dépôt après le retour",
)

data class GeofenceDecision(
    val eventType: GeofenceEventType,
    val nextSessionState: SessionState,
    /** Faut-il marquer le retour (horodatage et position) ? */
    val markReturned: Boolean,
    val alert: GeofenceAlert?,
    /** Explication lisible, conservée dans l'événement pour le support. */
    val reason: String,
)

fun evaluateGeofenceTransition(
    transition: TransitionKind,
    occurredAt: Instant,
    schedule: DepotSchedule,
    sessionState: SessionState,
): GeofenceDecision {
    if (transition == TransitionKind.ENTER) {
        if (!schedule.isAfterReturnTime(occurredAt)) {
            // Passage au dépôt en cours de tournée : ce n'est pas une fin de journée.
            return GeofenceDecision(
                eventType = GeofenceEventType.ENTER_DEPOT,
                nextSessionState = sessionState,
                markReturned = false,
                alert = null,
                reason = "Entrée au dépôt avant l'heure de retour : aucun effet.",
            )
        }

        return GeofenceDecision(
            eventType = GeofenceEventType.ENTER_DEPOT_AFTER_RETURN_TIME,
            nextSessionState = SessionState.RETURNED,
            // L'horodatage de retour n'est pas réécrit si la session est déjà
            // marquée retournée : le premier retour fait foi.
            markReturned = sessionState != SessionState.RETURNED,
            alert = null,
            reason = "Entrée au dépôt après l'heure de retour : téléphone considéré comme retourné.",
        )
    }

    if (sessionState == SessionState.RETURNED) {
        return GeofenceDecision(
            eventType = GeofenceEventType.AFTER_RETURN_EXIT,
            // L'alerte n'est pas un état bloquant : la session redevient active
            // et l'alerte reste ouverte jusqu'à acquittement (docs/02 §6).
            nextSessionState = SessionState.ACTIVE,
            markReturned = false,
            alert = GeofenceAlert(),
            reason = "Sortie du dépôt alors que le téléphone était marqué comme retourné.",
        )
    }

    return GeofenceDecision(
        eventType = GeofenceEventType.EXIT_DEPOT,
        nextSessionState = sessionState,
        markReturned = false,
        alert = null,
        reason = "Sortie du dépôt en cours de tournée : aucun effet.",
    )
}

// ---------------------------------------------------------------------------
//  Classification géométrique
// ---------------------------------------------------------------------------

enum class FixClassification { INSIDE_CERTAIN, OUTSIDE_CERTAIN, UNDETERMINED }

private const val EARTH_RADIUS_M = 6_371_008.8

/** Distance haversine, en mètres. */
fun haversineMeters(
    lat1: Double,
    lon1: Double,
    lat2: Double,
    lon2: Double,
): Double {
    val toRad = { degrees: Double -> degrees * PI / 180.0 }
    val dLat = toRad(lat2 - lat1)
    val dLon = toRad(lon2 - lon1)
    val a = sin(dLat / 2).let { it * it } +
        cos(toRad(lat1)) * cos(toRad(lat2)) * sin(dLon / 2).let { it * it }
    return 2 * EARTH_RADIUS_M * asin(min(1.0, sqrt(a)))
}

/**
 * Classe un point par rapport à une zone, en tenant compte de son incertitude.
 *
 *   d + a  <  R        -> certainement dedans
 *   d - a  >  R + H    -> certainement dehors
 *   sinon              -> indéterminé, AUCUNE transition
 *
 * C'est cette troisième branche qui absorbe le faux positif décrit dans la
 * spécification : 50 m hors zone avec 100 m de précision reste indéterminé.
 */
fun classifyFix(
    distanceMeters: Double,
    accuracyMeters: Double,
    radiusMeters: Double,
    hysteresisMeters: Double,
): FixClassification = when {
    distanceMeters + accuracyMeters < radiusMeters -> FixClassification.INSIDE_CERTAIN
    distanceMeters - accuracyMeters > radiusMeters + hysteresisMeters ->
        FixClassification.OUTSIDE_CERTAIN
    else -> FixClassification.UNDETERMINED
}
